smithy-aws-cloudformation-traits/src/main/java/software/amazon/smithy/aws/cloudformation/traits/CfnResource.java

```diff
@@ -0,0 +1,254 @@
+/*
+ * Copyright 2020 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License").
+ * You may not use this file except in compliance with the License.
+ * A copy of the License is located at
+ *
+ *  http://aws.amazon.com/apache2.0
+ *
+ * or in the "license" file accompanying this file. This file is distributed
+ * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+ * express or implied. See the License for the specific language governing
+ * permissions and limitations under the License.
+ */
+
+package software.amazon.smithy.aws.cloudformation.traits;
+
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import java.util.Set;
+import java.util.function.Function;
+import java.util.function.Predicate;
+import java.util.stream.Collectors;
+import software.amazon.smithy.model.shapes.ShapeId;
+import software.amazon.smithy.utils.MapUtils;
+import software.amazon.smithy.utils.SmithyBuilder;
+import software.amazon.smithy.utils.ToSmithyBuilder;
+
+/**
+ * Contains extracted resource information.
+ */
+public final class CfnResource implements ToSmithyBuilder<CfnResource> {
+    private final Map<String, CfnResourceProperty> propertyDefinitions = new HashMap<>();
+    private final Set<ShapeId> excludedProperties = new HashSet<>();
+    private final Set<String> primaryIdentifiers = new HashSet<>();
+    private final List<Set<String>> additionalIdentifiers = new ArrayList<>();
+
+    private CfnResource(Builder builder) {
+        this.propertyDefinitions.putAll(builder.propertyDefinitions);
+        this.excludedProperties.addAll(builder.excludedProperties);
+        this.primaryIdentifiers.addAll(builder.primaryIdentifiers);
+        this.additionalIdentifiers.addAll(builder.additionalIdentifiers);
+    }
+
+    public static Builder builder() {
+        return new Builder();
+    }
+
+    /**
+     * Get all property definitions of the CloudFormation resource.
+     *
+     * @return Returns all members that map to CloudFormation resource
+     *   properties.
+     */
+    public Map<String, CfnResourceProperty> getProperties() {
+        return propertyDefinitions.entrySet().stream()
```

Comment:
Can this just be computed up front? It gets called a lot, and it might make the implementation clearer IMO.